ENH: Add transform analysis utils

jhlegarreta opened the pull request:
Add transform analysis utils.

Transfer contents from the `NiFreeze` projects so that hey can be reused across projects requiring transform analysis:
https://github.com/nipreps/nifreeze/blob/d27ba7552bbd9095c3c13b46443d87a4b5504c4c/src/nifreeze/analysis/motion.py https://github.com/nipreps/nifreeze/blob/d27ba7552bbd9095c3c13b46443d87a4b5504c4c/src/nifreeze/data/utils.py

Add a fixture to be able to reuse a random number generator across tests.

Fixes #239.

jhlegarreta committed:
ENH: Add transform analysis utils

Add transform analysis utils.

Transfer contents from the `NiFreeze` projects so that hey can be reused
across projects requiring transform analysis:
https://github.com/nipreps/nifreeze/blob/d27ba7552bbd9095c3c13b46443d87a4b5504c4c/src/nifreeze/analysis/motion.py
https://github.com/nipreps/nifreeze/blob/d27ba7552bbd9095c3c13b46443d87a4b5504c4c/src/nifreeze/data/utils.py

Add a fixture to be able to reuse a random number generator across
tests.

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

jhlegarreta commented on nitransforms/utils.py:
Removed the file.

oesteban reviewed (commented):
Thanks for the update. I just realized of a little inconsistency. Other than that, I think this is almost ready.

jhlegarreta commented:
This is ready to go. Merging this would allow `NiFreeze` to drop code that is general, and probably have better support as these functions may get better testing. We will need to set/bump a new min version for `NiTransforms` when a new version with these changes is released.

oesteban commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,188 @@
+
+
+RADIUS = 50.0
+"""Typical radius (in mm) of a sphere mimicking the size of a typical human brain."""
```
```suggestion
"""
Default radius (in mm) of a sphere where framewise displacements are calculated.
The choice was proposed by
`Power et al. (2015) <https://doi.org/10.1016/j.neuroimage.2011.10.018>`__, and it
represents approximately the mean distance from the cerebral cortex to the center of the head.
"""
```

oesteban commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,188 @@
+from nitransforms.base import TransformBase
+
+
+RADIUS = 50.0
```
```suggestion
DEFAULT_FD_RADIUS = 50.0
```

oesteban commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,188 @@
+"""Typical radius (in mm) of a sphere mimicking the size of a typical human brain."""
+
+
+def compute_fd_from_motion(motion_parameters: np.ndarray, radius: float = RADIUS) -> np.ndarray:
```
```suggestion
def compute_fd_from_motion(motion_parameters: np.ndarray, radius: float = DEFAULT_FD_RADIUS) -> np.ndarray:
```

oesteban commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,188 @@
+
+
+def compute_fd_from_motion(motion_parameters: np.ndarray, radius: float = RADIUS) -> np.ndarray:
+    """Compute framewise displacement (FD) from motion parameters.
```
We should cite [Power et al. (2013)](https://doi.org/10.1016/j.neuroimage.2011.10.018) (doi:10.1016/j.neuroimage.2011.10.018) somewhere here.

jhlegarreta replied:
Done in 883bf93.

oesteban commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,188 @@
+    return *translation, *rotation_deg
+
+
+def identify_spikes(fd: np.ndarray, threshold: float = 2.0) -> Tuple[np.ndarray, np.ndarray]:
```
Not sure this is something very useful that we should maintain. In any case:

```suggestion
def identify_spikes(fd: np.ndarray, z_threshold: float = 2.0) -> Tuple[np.ndarray, np.ndarray]:
```

oesteban commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,188 @@
+    # Normalize (z-score)
+    fd_norm = zscore(fd)
+
+    mask = fd_norm > threshold
```
```suggestion
    mask = fd_norm > z_threshold
```

oesteban commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,188 @@
+    # Convert to world coordinates
+    center_xyz = nb.affines.apply_affine(affine, center_ijk)
+    # Generate coordinates of points at radius distance from center
+    fd_coords = np.array(list(product(*((radius, -radius),) * 3))) + center_xyz
```
BUG: the radius of these points is not 50 mm:

```
In [1]: from itertools import product
    ...: import numpy as np
    ...: 
    ...: radius = 50
    ...: center_xyz = (0, 0, 0)
    ...: 
    ...: fd_coords = np.array(list(product(*((radius, -radius),) * 3))) + center_xyz
    ...: np.linalg.norm(fd_coords, axis=1)
Out[1]: 
array([86.60254038, 86.60254038, 86.60254038, 86.60254038, 86.60254038,
       86.60254038, 86.60254038, 86.60254038])
```

cc/ @effigies in the likely case I brought over this bug from other places (nipype?, fmriprep?)

oesteban replied:
```suggestion
    e_coords = np.array(list(product(*((1, -1),) * 3)))
    e_coords = e_coords / np.linalg.norm(e_coords, axis=1)[:, None]
    fd_coords = np.vstack((np.eye(3), -1 * np.eye(3), e_xyz)) * radius + center_xyz
```

We should extract this into a function (I think DIPY has something like this?) that we can test thoroughly:

```Python
import math
import numpy as np
from itertools import product


def sample_unit_sphere(n_points=8):
    """Returns :math:`N` evenly distributed points on a unit-radius sphere.

    This function returns a **deterministic**, **quasi-uniform** point set on the
    surface of the unit sphere :math:`S^2 \\subset \\mathbb{R}^3`.

    Notes
    -----
    - There is no unique notion of "evenly distributed" for arbitrary `N` on a sphere.
      This function uses:
        * **Platonic solids** for certain small `N` (high symmetry; e.g. `N=6` gives
          the ±axis points).
        * A **Fibonacci / golden-angle spiral** otherwise (fast, simple, good coverage).

    Parameters
    ----------
    n_points : int
        Number of points on the sphere.

    Returns
    -------
    numpy.ndarray
        An array of shape ``(n_points, 3)`` whose rows have unit norm.

    Examples
    --------
    Basic shape + unit norm:

    >>> import numpy as np
    >>> X = sample_unit_sphere(10)
    >>> X.shape
    (10, 3)
    >>> np.allclose(np.linalg.norm(X, axis=1), 1.0)
    True

    Edge case:

    >>> sample_unit_sphere(0)
    Traceback (most recent call last):
    ...
    ValueError: n_points must be a positive integer

    For N=6, return the ±axis points (octahedron vertices), matching your example idea:

    >>> X = sample_unit_sphere(6)
    >>> # Each row has exactly one coordinate with magnitude 1, others 0
    >>> np.all((np.abs(X) == 1.0).sum(axis=1) == 1)
    True
    >>> np.all((np.abs(X) == 1.0).sum(axis=0) == 2)  # each axis appears twice (±)
    True

    For N=4, the tetrahedron has constant pairwise dot product -1/3 off-diagonal:

    >>> X = sample_unit_sphere(4)
    >>> D = X @ X.T
    >>> off = D[~np.eye(4, dtype=bool)]
    >>> np.allclose(off, -1/3)
    True

    For a quasi-uniform set, the second moment matrix is close to I/3, and the
    minimum angular separation is non-trivial:

    >>> X = sample_unit_sphere(200)
    >>> M = (X.T @ X) / len(X)
    >>> np.allclose(M, np.eye(3) / 3, atol=1e-3)
    True
    >>> def min_angle_rad(Y):
    ...     dots = np.clip(Y @ Y.T, -1.0, 1.0)
    ...     np.fill_diagonal(dots, 1.0)
    ...     ang = np.arccos(dots)
    ...     np.fill_diagonal(ang, np.inf)
    ...     return float(ang.min())
    >>> min_angle_rad(X) > 0.18
    True
    """
    if isinstance(n_points, (bool, np.bool_)) or not isinstance(n_points, (int, np.integer)):
        raise TypeError("n_points must be an integer")
    if n_points < 1:
        raise ValueError("n_points must be a positive integer")

    def _normalize(X):
        X = np.asarray(X, dtype=float)
        X /= np.linalg.norm(X, axis=1, keepdims=True)
        return X

    # --- Highly symmetric small-N cases (Platonic solids / degeneracies) ---
    if n_points == 1:
        return np.array([[0.0, 0.0, 1.0]])
    if n_points == 2:
        return np.array([[0.0, 0.0, 1.0],
                         [0.0, 0.0, -1.0]])
    if n_points == 4:  # tetrahedron (4 cube corners with even number of minus signs)
        X = np.array([[ 1.0,  1.0,  1.0],
                      [ 1.0, -1.0, -1.0],
                      [-1.0,  1.0, -1.0],
                      [-1.0, -1.0,  1.0]])
        return _normalize(X)
    if n_points == 6:  # octahedron: ±axes
        return np.vstack([np.eye(3), -np.eye(3)]).astype(float)
    if n_points in (8, 20):  # hexahedron (cube) & base for dodecahedron
        X = np.array(list(product([-1.0, 1.0], repeat=3)), dtype=float)
        if n_points == 8:
            return _normalize(X)

        # Dodecahedron, add 12 vertices
        extra = []
        phi = (1.0 + math.sqrt(5.0)) / 2.0
        invphi = 1.0 / phi
        for a in (-1.0, 1.0):
            for b in (-1.0, 1.0):
                extra.extend([
                    [0.0, a * invphi, b * phi],
                    [a * invphi, b * phi, 0.0],
                    [a * phi, 0.0, b * invphi],
                ])
        X = np.vstack([X, np.asarray(extra, dtype=float)])
        return _normalize(X)
    if n_points == 12:  # icosahedron
        phi = (1.0 + math.sqrt(5.0)) / 2.0
        X = []
        for a in (-1.0, 1.0):
            for b in (-1.0, 1.0):
                X.append([0.0, a, b * phi])
                X.append([a, b * phi, 0.0])
                X.append([a * phi, 0.0, b])
        return _normalize(X)

    # --- General N: Fibonacci / golden-angle spiral (deterministic quasi-uniform) ---
    i = np.arange(n_points, dtype=float)
    golden_angle = math.pi * (3.0 - math.sqrt(5.0))

    z = 1.0 - 2.0 * (i + 0.5) / n_points
    r = np.sqrt(np.maximum(0.0, 1.0 - z * z))
    theta = golden_angle * i

    X = np.column_stack((r * np.cos(theta), r * np.sin(theta), z))
    return _normalize(X)

```


EDIT: Added a suggestion of how we could implement this here. This function could also be useful for surface representations in nibabel.

jhlegarreta replied:
There are a number of ways to create the vertices over the sphere in DIPY:
https://github.com/dipy/dipy/blob/8009c8124ee7932701a375ab105373350a5c5a38/dipy/core/sphere.py
Most notably,
https://github.com/dipy/dipy/blob/8009c8124ee7932701a375ab105373350a5c5a38/dipy/core/sphere.py#L482
https://github.com/dipy/dipy/blob/8009c8124ee7932701a375ab105373350a5c5a38/dipy/core/sphere.py#L542

oesteban reviewed (commented):
We had a pretty hefty bug in the original implementation of nifreeze!

jhlegarreta commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,188 @@
+    ----------
+    fd : :obj:`~numpy.ndarray`
+        Framewise displacement data.
+    threshold : :obj:`float`, optional
```
```suggestion
    z_threshold : :obj:`float`, optional
```

jhlegarreta commented on nitransforms/tests/test_analysis.py:
```diff
@@ -0,0 +1,156 @@
+    n_samples = 450
+
+    fd = rng.normal(0, 5, n_samples)
+    threshold = 2.0
```
```suggestion
    z_threshold = 2.0
```

jhlegarreta commented on nitransforms/tests/test_analysis.py:
```diff
@@ -0,0 +1,156 @@
+    expected_mask = np.zeros(n_samples, dtype=bool)
+    expected_mask[expected_indices] = True
+
+    obtained_indices, obtained_mask = identify_spikes(fd, threshold=threshold)
```
```suggestion
    obtained_indices, obtained_mask = identify_spikes(fd, z_threshold=z_threshold)
```

jhlegarreta committed:
Apply suggestions from code review

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

jhlegarreta committed:
DOC: Cite Power et al. in FD computation function docstring

Cite Power et al. in FD computation function docstring.

jhlegarreta commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,197 @@
+def compute_fd_from_transform(
+    img: nb.spatialimages.SpatialImage,
+    test_xfm: TransformBase,
+    radius: float = RADIUS,
```
```suggestion
    radius: float = DEFAULT_FD_RADIUS,
```

jhlegarreta committed:
Update nitransforms/analysis/utils.py

jhlegarreta committed:
STY: Fix style errors

Fix style errors.

oesteban committed:
fix: generalize sampling on unit sphere computation

oesteban committed:
fix: defensive increase of tolerance in FD test

oesteban committed:
doc: add missing modules to apidoc

oesteban committed:
tst: exercise all branches

oesteban committed:
fix: address bug in ``compute_fd_from_transform``

oesteban commented:
@jhlegarreta please have one more look - I realized that the FD from transform matrices was also badly computed as it did not accept the previous transform. IMHO, we should also provide a "vectorized" version of FD such that you pass the list of transforms and gives you timeseries. Could be left for a future PR though.

@effigies added you to have an extra pair of eyes on this.

oesteban commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,376 @@
+    xfm: TransformBase,
+    xfm_prev: TransformBase | None = None,
+    radius: float = DEFAULT_FD_RADIUS,
+    n_vertices: int = 8,
```
It would be interesting to make the output match ``compute_fd_from_motion`` for ``n_vertices = 1``.

jhlegarreta replied:
See cbc29d481346610468bf5da2c3cb9fd5e3dd951b.

jhlegarreta commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,376 @@
+    Compute the framewise displacement (FD) for a given transformation.
+
+    This implementation varies with respect to the original formulation by [Power2012]_
+    in that the FD is computed as the average across a number of vertices sample of the
```
```suggestion
    in that the FD is computed as the average across a number of vertices sampled over the
```

jhlegarreta replied:
The method to sample the sphere should be documented or cross referenced using Sphinx markup in the docstring.

jhlegarreta commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,376 @@
+
+    Notes
+    -----
+    - There is no unique notion of "evenly distributed" for arbitrary `N` on a sphere.
```
```suggestion
    - There is no unique notion of "evenly distributed" for arbitrary ``N`` on a sphere.
```

jhlegarreta commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,376 @@
+    -----
+    - There is no unique notion of "evenly distributed" for arbitrary `N` on a sphere.
+      This function uses:
+        * **Platonic solids** for certain small `N` (high symmetry; e.g. `N=6` gives
```
```suggestion
        * **Platonic solids** for certain small ``N`` (high symmetry; e.g. ``N=6`` gives
```

jhlegarreta commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,376 @@
+    ((12, 3), True)
+    ((20, 3), True)
+
+    For N=6, return the ±axis points (octahedron vertices):
```
```suggestion
    For ``N=6``, return the ±axis points (octahedron vertices):
```

jhlegarreta commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,376 @@
+    >>> bool(np.all((np.abs(X) == 1.0).sum(axis=0) == 2))  # each axis appears twice (±)
+    True
+
+    For N=4, the tetrahedron has constant pairwise dot product -1/3 off-diagonal:
```
```suggestion
    For ``N=4``, the tetrahedron has constant pairwise dot product -1/3 off-diagonal:
```

jhlegarreta commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,376 @@
+    >>> bool(np.allclose(off, -1/3))
+    True
+
+    For a quasi-uniform set, the second moment matrix is close to I/3, and the
```
```suggestion
    For a quasi-uniform set, the second moment matrix is close to ``I/3``, and the
```

jhlegarreta reviewed (commented):
> IMHO, we should also provide a "vectorized" version of FD such that you pass the list of transforms and gives you timeseries. Could be left for a future PR though.

Better for a future PR. This is already better than what we have in `NiFreeze` (thanks for reviewing and the subsequent work), allows us to have a common, single authoritative location, and frees `NiFreeze` from hosting this.

Some inline comments about documentation and syntax highlighting.

jhlegarreta commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,376 @@
+
+    This implementation varies with respect to the original formulation by [Power2012]_
+    in that the FD is computed as the average across a number of vertices sample of the
+    sphere.
```
```suggestion
    sphere. See :func:`~nitransforms.analysis.utils.sample_unit_sphere` for details
    about the vertex sampling method.
```

jhlegarreta committed:
Apply suggestions from code review

jhlegarreta committed:
Apply suggestions from code review

effigies commented:
Out of curiosity, why are we standardizing on degrees? Apart from AFNI, everybody reports angles in radians. It seems to add unnecessary complication. Nipype also normalizes to SPM style (https://github.com/nipy/nipype/blob/8234ec318489930fa17487cf15d210420214b00a/nipype/utils/misc.py#L246-L271) so consistency within the nipy world seems desirable.

effigies commented:
Another conceptual question: Do we want to continue the tradition of accepting an unlabeled Nx6 matrix, or would it be a good opportunity to require the caller to consider the  it useful to have a function that accepts an Nx6 matrix, or would it make more sense to accept 2 Nx3 matrices and a radius? The reason I ask is that it would be very tempting to load an Nx6 matrix out of some transform file and pass it directly without carefully considering which are the translations and which are the rotations. I might think of the API as follows:

```py
def framewise_displacement(
    translations: np.ndarray,
    rotations: np.ndarray,
    radius: float = DEFAULT_RADIUS
) -> np.ndarray:
    """Calculate framewise displacement of motion parameters. ..."""

def extract_parameters(motion_parameters: np.ndarray, fmt: str | None=None) -> tuple[np.ndarray, np.ndarray]:
    """Extract translation and rotation parameters from a parameter array.

    This function normalizes rotation parameters to radians, if the input format is known to use degrees.

    The parameters are returned in the order expected by the `framewise_displacement` function.
    """

# I know what my matrix is:
fd = framewise_displacement(motion_parameters[:, :3], motion_parameters[:, 3:])
# I know I got it from AFNI:
afni_fd = framewise_displacement(*extract_parameters(motion_parameters, fmt='afni'))
```

We could even force people to use labels instead and say:

```py
@dataclass(kw_only=True)
class MotionParameters:
    rotations: np.ndarray
    translations: np.ndarray

def extract_parameters(motion_parameters: np.ndarray, fmt: str | None=None) -> MotionParameters: ...
def framewise_displacement(motion_parameters: MotionParameters, radius: float = DEFAULT_RADIUS) -> np.ndarray: ...
```

This would even suggest a straightforward translation between affines and motion parameters:

```py
def affineToMotionParams(aff: Affine) -> MotionParameters: ...
def motionParamstoAffine(motion_parameters: MotionParameters) -> LinearTransformsMapping: ...
```

Just a couple thoughts. Will review the PR on the assumption that you'll stick with degrees and a big matrix.

effigies commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,377 @@
+    rotation_displacement = d_rotations * radius
+
+    # Compute FD as sum of absolute differences
+    return np.sum(np.abs(d_translations) + np.abs(rotation_displacement), axis=1)
```
Conceptually, this last line reads weirdly to me. Why add the translation displacements to the rotation displacements and then sum? It does the job, but from the perspective of understanding the method, I might write this as:

```suggestion
    translations = motion_parameters[:, :3]
    rotations = np.deg2rad(motion_parameters[:, 3:])
    
    displacements = np.hstack((
        np.diff(translations, axis=0, prepend=0),
        np.diff(rotations * radius, axis=0, prepend=0)
    ))

    # FD is the L1 norm (sum of absolute values)
    return np.linalg.norm(displacements, ord=1, axis=1)
```

effigies commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,377 @@
+    """
+    Compute the framewise displacement (FD) for a given transformation.
+
+    This implementation varies with respect to the original formulation by [Power2012]_
+    in that the FD is computed as the average across a number of vertices sampled over the
+    sphere. See :func:`~nitransforms.analysis.utils.sample_unit_sphere` for details
+    about the vertex sampling method.
```
This also uses the L2 norm instead of L1. What is the reasoning for this?

oesteban replied:
Thanks for pointing this out, changing to L1

effigies commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,377 @@
+    return np.linalg.norm(diffs, axis=-1)
+
+
+def extract_motion_parameters(affine: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+    """Extract translation (mm) and rotation (degrees) parameters from an affine matrix.
+
+    Parameters
+    ----------
+    affine : :obj:`~numpy.ndarray`
+        The affine transformation matrix.
+
+    Returns
+    -------
+    :obj:`tuple`
+        Extracted translation and rotation parameters.
+    """
+
+    translation = affine[:3, 3]
+    rotation_rad = np.arctan2(
+        [affine[2, 1], affine[0, 2], affine[1, 0]],
+        [affine[2, 2], affine[0, 0], affine[1, 1]],
+    )
+    rotation_deg = np.rad2deg(rotation_rad)
+    return *translation, *rotation_deg
```
This isn't obviously correct to me. There are many ways of extracting angles from a rotation matrix, but I'm surprised that you seem able to get them using only 6 entries of a 3x3 matrix.

At the least, this function needs a citation, but I would be inclined to rely on scipy.spatial.transforms or transforms3d for this functionality.

oesteban replied:
Thanks, will investigate.

jhlegarreta replied:
See 18faeb80725e9ec48eba978c4f16db7503cff954.

effigies commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,377 @@
+    numpy.ndarray
+        An array of shape ``(n_points, 3)`` whose rows have unit norm.
+
+    Examples
```
Would it be too much to ask for these docs to have plots to show the points on the unit sphere?

oesteban replied:
Sounds good, I'll see if it doesn't involve much or defer to a future PR.

jhlegarreta replied:
See b29f8398c98936ec9b6675e80ca91057aabc5201.

effigies commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,377 @@
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Identify motion spikes in framewise displacement data.
+
+    Identifies high-motion frames as timepoint exceeding a given threshold value
+    based on z-score normalized framewise displacement (FD) values.
```
Is there a citation for this one? This seems dubious to me, as z-scoring will amplify low-motion subjects' FD and suppress high-motion subjects' FD. For the purposes of censoring, absolute thresholds are often used based on voxel size.

oesteban replied:
I'm inclined to drop this function altogether. The purpose is not censoring, but rather identifying the higher trees in the forest.

jhlegarreta replied:
Removed in b459594b89f8fc5e5233e784b17634a45eacc3c7.

oesteban commented on nitransforms/analysis/utils.py:
```diff
@@ -0,0 +1,377 @@
+    # Generate coordinates of points at radius distance from center
+    fd_coords = sample_unit_sphere(n_points=n_vertices) * radius + center_xyz
+    # Compute the average displacement from the test transformation
+    return np.mean(np.linalg.norm(xfm.map(fd_coords) - xfm_prev.map(fd_coords), axis=-1))
```
```suggestion
    return np.mean(np.linalg.norm(xfm.map(fd_coords) - xfm_prev.map(fd_coords), ord=1, axis=-1))
```

oesteban committed:
Apply suggestions from code review

Co-authored-by: Chris Markiewicz <effigies@gmail.com>

jhlegarreta commented:
Can we revisit this so that we can reduce the fronts across NiFreeze, please?

jhlegarreta committed:
REF: Remove spike identification function

Remove spike identification function.

jhlegarreta committed:
REF: Use `scipy` to extract rotations from affine matrices

Use `scipy` to extract rotations from affine matrices.

jhlegarreta committed:
REF: Make the computation of FD for a single vertex consistent

Make the computation of FD for a single vertex consistent.

jhlegarreta committed:
REF: Ensure motion array has right dimensionality before computing FD

Ensure motion array has right dimensionality before computing FD.

Take advantage of the commit to:
- Make the array use floating point precision.
- Prefer using explicit array rather than relying on broadcasting for
  prepending.

jhlegarreta committed:
DOC: Improve return value docstring in FD from motion function

Improve return value docstring in FD computation function from motion
array.

jhlegarreta committed:
DOC: Plot unit sphere sampling docstring examples

Plot unit sphere sampling docstring examples.

Take advantage of the commit to:
- Define the values used for the docstring test as a named variable for
  the sake of clearliness.
- Use Sphinx markup for math symbols instead of using non-unicode
  characters.

jhlegarreta commented:
I think https://github.com/nipy/nitransforms/pull/286#issuecomment-3719915982 can be addressed in a subsequent PR.

jhlegarreta commented:
I'd squash commits once we are satisfied with the changes.

jhlegarreta committed:
TST: Use L1 norm to check FD value from transform matrix

Use L1 norm to check resulting FD value from transform matrix.

jhlegarreta commented:
Note: we are assuming that the matrices are all valid: scipy 1.17 introduced a [check parameter](https://docs.scipy.org/doc/scipy-1.17.0/reference/generated/scipy.spatial.transform.Rotation.from_matrix.html) that we are not able to use [because we only require scipy >= 1.10](https://docs.scipy.org/doc/scipy-1.16.2/reference/generated/scipy.spatial.transform.Rotation.from_matrix.html).

I think I've done all I can on this PR.

jhlegarreta committed:
DOC: Improve documentation

Improve documentation: use appropriate markup for types, document
exceptions.

jhlegarreta committed:
TST: Test exceptions

Test exceptions.

jhlegarreta commented:
@oesteban @effigies Can you please look at this so that we can make another step forward with the bits that we'd like to outsource to here in NiFreeze?

oesteban commented:
I'm looking at the RTD documentation and, in addition to see an outdated version (?) showing [``:math:`pm`axis points``](https://nitransforms--286.org.readthedocs.build/en/286/_api/analysis.html#id4), I am not seeing any pictures showing where points are located, am I looking at the wrong place?

jhlegarreta commented:
> I'm looking at the RTD documentation and, in addition to see an outdated version (?) showing [:math:`pm`axis points](https://nitransforms--286.org.readthedocs.build/en/286/_api/analysis.html#id4), I am not seeing any pictures showing where points are located, am I looking at the wrong place?

Math symbols have been fixed and the plots are shown now.
